### PR TITLE
Sanitize query property names in JSON/YAML.mapping

### LINKED ISSUE #4129: JSON.mapping allow special getter? for Bool
Given this JSON file:
```json
{
  "id": 42,
  "name": "bew",
  "is_bot": false
}
```
The corresponding mapping would be:
```crystal
class User
  JSON.mapping(
    id: Int32,
    name: String,
    bot: {type: Bool, key: "is_bot"},
  )
end
```
This would allow us to write:
```crystal
user = User.from_json some_json
p user.name # => bew
p user.bot # => false
```
The issue here is the `bot` which is not really crystal-ish.. I want my method to be `bot?` not just `bot`

Maybe we could do something like:
```crystal
JSON.mapping(
  bot: {type: Bool, getter?: true},     # note the ? on getter
)
```
This would result in a getter with `?` (should be the default for `Bool` ?)
```crystal
p user.bot? # => false
```

What do you think ?

### PR DESCRIPTION
**Pull request by Sija**
Closes #4129

### CODE REVIEW

**Comment by bew** [comment]
Can you also add support to `getter!` as in `{foo!: "bar"}` ?

**Comment by Sija** [comment]
@bew no, I was thinking about it, but haven't found valid usecases.

**Comment by Sija** [comment]
@bcardiff @RX14 @bew Could you please spare a review?

**Comment by Sija** [comment]
Will it merge? :)

**Comment by bew** [comment]
I'm not sure I like this `safe_key` field in value. We already have `key` field in value, which represent the JSON/YAML key name. To me `safe_key` is not explicit that it is the instance's variable name.

I know it's an implementation detail, but from reading the code it's not easy to understand what this `safe_key` field is for.

**Comment by Sija** [comment]
@bew I'm not fond of it either, feel free to propose better name :) btw, it doesn't only resprent ivar name, since it's used for example in `<key>_present?` query methods.

**Comment by Sija** [comment]
Last recommended action was changing key name from `:safe_key` to sth more apt by @bew, yet no proposals for a better name was posted. Is it GTG as is then?

**Comment by straight-shoota** [comment]
What about `key_id`? It essentially replaces `key.id`...

**Comment by Sija** [comment]
@straight-shoota 👍  changed, thx.

**Comment by Sija** [comment]
@jhass @mverzilli @ysbaddaden plz review ❤️

**Comment by asterite** [comment]
I understand that `foo?:` parses, just like `foo? = 1`, but that doesn't parse in Ruby and it looks weird. I'm not sure it should parse... that said, maybe it doesn't hurt, but I don't know.

**Comment by Sija** [comment]
@asterite setter is generated without `?` suffix, so `foo?:` => `obj.foo = 1`.

**Comment by asterite** [comment]
@Sija  I mean, `foo?:` shouldn't parse. It doesn't in Ruby.

**Comment by Sija** [comment]
@asterite I misunderstood you. I think it's a useful feature, otherwise I see no simple solution to the issue at hand.

**Comment by Sija** [comment]
@asterite If `foo?: 1` syntax stays within the language (which I'd strongly prefer), this PR can be merged without any side-effects, while solving common use case.

**Comment by asterite** [comment]
Yeah, I think that syntax is ok. More consistent. In fact maybe we could use it for property, getter, etc, instead of the uggly getter? and property? macros. I'll play a bit with that.

File: src/json/mapping.cr [hunk 1]
```diff
@@ -67,25 +67,29 @@ module JSON
     {% end %}
 
     {% for key, value in _properties_ %}
-      @{{key.id}} : {{value[:type]}} {{ (value[:nilable] ? "?" : "").id }}
+      {% _properties_[key][:key_id] = key.id.gsub(/\?$/, "") %}
+    {% end %}
+
+    {% for key, value in _properties_ %}
+      @{{value[:key_id]}} : {{value[:type]}} {{ (value[:nilable] ? "?" : "").id }}
```
Comment (RX14) [inline src/json/mapping.cr]: Would `ivar_name` be a better name than `key_id` here?

**Pull request merged**